packages/firestore/src/model/values.ts

```diff
@@ -556,26 +560,36 @@ export function isNumber(value?: api.Value | null): boolean {
 }
 
 /** Returns true if `value` is an ArrayValue. */
-export function isArray(value?: api.Value | null): boolean {
+export function isArray(
+  value?: api.Value | null
+): value is { arrayValue: api.ArrayValue } {
   return !!value && 'arrayValue' in value;
 }
 
 /** Returns true if `value` is a ReferenceValue. */
-export function isReferenceValue(value?: api.Value | null): boolean {
+export function isReferenceValue(
+  value?: api.Value | null
+): value is { referenceValue: string } {
   return !!value && 'referenceValue' in value;
 }
 
 /** Returns true if `value` is a NullValue. */
-export function isNullValue(value?: api.Value | null): boolean {
+export function isNullValue(
+  value?: api.Value | null
+): value is { nullValue: 'NULL_VALUE' } {
   return !!value && 'nullValue' in value;
 }
 
 /** Returns true if `value` is NaN. */
-export function isNanValue(value?: api.Value | null): boolean {
-  return !!value && isNaN(Number(value.doubleValue));
+export function isNanValue(
+  value?: api.Value | null
+): value is { doubleValue: 'NaN' | number } {
+  return !!value && isDouble(value) && isNaN(Number(value.doubleValue));
```

Comment:
`isDouble` already tests `!!value` so you can drop it here.

Reply:
Done